File: solo-leveling-player-service/src/main/java/com/sleepkqq/sololeveling/player/service/kafka/consumer/SaveTasksConsumer.java
package com.sleepkqq.sololeveling.player.service.kafka.consumer;

import static com.sleepkqq.sololeveling.avro.constants.KafkaGroupIds.PLAYER_GROUP_ID;
import static com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics.SAVE_TASKS_TOPIC;
import static com.sleepkqq.sololeveling.avro.notification.NotificationPriority.LOW;
import static com.sleepkqq.sololeveling.player.service.model.player.enums.PlayerTaskStatus.IN_PROGRESS;

import com.sleepkqq.sololeveling.avro.notification.Notification;
import com.sleepkqq.sololeveling.avro.notification.SendNotificationEvent;
import com.sleepkqq.sololeveling.avro.task.SaveTasksEvent;
import com.sleepkqq.sololeveling.player.service.kafka.producer.SendNotificationProducer;
import com.sleepkqq.sololeveling.player.service.mapper.AvroMapper;
import com.sleepkqq.sololeveling.player.service.model.Immutables;
import com.sleepkqq.sololeveling.player.service.service.player.PlayerTaskService;
import com.sleepkqq.sololeveling.player.service.service.task.TaskService;
import java.time.LocalDateTime;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import one.util.streamex.StreamEx;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
@RequiredArgsConstructor
public class SaveTasksConsumer {

  private final TaskService taskService;
  private final PlayerTaskService playerTaskService;
  private final SendNotificationProducer sendNotificationProducer;
  private final AvroMapper avroMapper;

  @KafkaListener(
      topics = SAVE_TASKS_TOPIC,
      groupId = PLAYER_GROUP_ID,
      containerFactory = "kafkaListenerContainerFactorySaveTasksEvent"
  )
  @Transactional
  public void listen(SaveTasksEvent event) {
    log.info(">> Start saving tasks | transactionId={}", event.getTransactionId());
    var now = LocalDateTime.now();
    StreamEx.of(event.getTasks())
        .map(t -> {
          var task = avroMapper.map(t);
          var version = taskService.getVersion(task.id());
          return taskService.update(
              Immutables.createTask(task, td -> td.setVersion(version)), now
          );
        })
        .forEach(t -> {
          var playerTaskId = playerTaskService.getTaskId(event.getPlayerId(), t.id());
          playerTaskService.setStatus(playerTaskId, IN_PROGRESS, now);
        });

    log.info("<< Tasks successfully saved | transactionId={}", event.getTransactionId());
    var sendNotificationEvent = new SendNotificationEvent(
        event.getTransactionId(),
        event.getPlayerId(),
        LOW,
        new Notification("Your tasks have been successfully generated!")
    );
    sendNotificationProducer.send(sendNotificationEvent);
  }
}

File: solo-leveling-player-service/src/main/kotlin/com/sleepkqq/sololeveling/player/service/kafka/consumer/SaveTasksConsumer.kt
package com.sleepkqq.sololeveling.player.service.kafka.consumer

import com.sleepkqq.sololeveling.avro.constants.KafkaGroupIds
import com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics
import com.sleepkqq.sololeveling.avro.notification.Notification
import com.sleepkqq.sololeveling.avro.notification.NotificationPriority
import com.sleepkqq.sololeveling.avro.notification.NotificationSource
import com.sleepkqq.sololeveling.avro.notification.NotificationType
import com.sleepkqq.sololeveling.avro.notification.SendNotificationEvent
import com.sleepkqq.sololeveling.avro.task.SaveTasksEvent
import com.sleepkqq.sololeveling.player.service.kafka.producer.SendNotificationProducer
import com.sleepkqq.sololeveling.player.service.lozalization.LocalizationCodes.TASKS_GENERATION_SUCCESS
import com.sleepkqq.sololeveling.player.service.mapper.AvroMapper
import com.sleepkqq.sololeveling.player.service.service.player.PlayerTaskService
import com.sleepkqq.sololeveling.player.service.service.player.PlayerTaskStatusService
import com.sleepkqq.sololeveling.player.service.service.redis.IdempotencyService
import com.sleepkqq.sololeveling.player.service.service.task.TaskService
import org.slf4j.LoggerFactory
import org.springframework.context.MessageSource
import org.springframework.context.i18n.LocaleContextHolder
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.support.Acknowledgment
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.Assert

@Suppress("unused")
@Service
class SaveTasksConsumer(
	private val taskService: TaskService,
	private val playerTaskService: PlayerTaskService,
	private val playerTaskStatusService: PlayerTaskStatusService,
	private val sendNotificationProducer: SendNotificationProducer,
	private val avroMapper: AvroMapper,
	private val idempotencyService: IdempotencyService,
	private val messageSource: MessageSource
) {

	private val log = LoggerFactory.getLogger(javaClass)

	@KafkaListener(
		topics = [KafkaTaskTopics.SAVE_TASKS_TOPIC],
		groupId = KafkaGroupIds.PLAYER_GROUP_ID,
		containerFactory = "kafkaListenerContainerFactorySaveTasksEvent"
	)
	@Transactional
	fun listen(event: SaveTasksEvent, ack: Acknowledgment) {
		val startTime = System.currentTimeMillis()
		val txId = event.transactionId

		log.info(">> Start processing SaveTasksEvent | transactionId={}", txId)

		try {
			if (idempotencyService.isProcessed(txId)) {
				ack.acknowledge()
				return
			}

			processSaveTasksEvent(event)

			sendSuccessNotification(event)

			val processingTime = System.currentTimeMillis() - startTime
			log.info(
				"<< Successfully processed SaveTasksEvent | transactionId={}, processingTime={}ms",
				txId, processingTime
			)

			ack.acknowledge()

		} catch (e: Exception) {
			val processingTime = System.currentTimeMillis() - startTime
			log.error(
				"Failed to process SaveTasksEvent | transactionId={}, processingTime={}ms, error={}",
				txId, processingTime, e.message, e
			)

			throw e
		}
	}

	private fun processSaveTasksEvent(event: SaveTasksEvent) {
		validateSaveTasksEvent(event)

		val tasks = event.tasks.map {
			avroMapper.map(it).toEntity()
		}

		log.info("Updating {} tasks for player {}", tasks.size, event.playerId)
		taskService.updateAll(tasks)

		val taskIds = tasks.map { it.id() }
		val playerTasks = playerTaskService.find(event.playerId, taskIds)

		if (playerTasks.isNotEmpty()) {
			log.info(
				"Setting {} player tasks to IN_PROGRESS for player {}",
				playerTasks.size,
				event.playerId
			)
			playerTaskStatusService.inProgressTasks(playerTasks)
		}
	}

	private fun validateSaveTasksEvent(event: SaveTasksEvent) {
		Assert.hasText(event.transactionId, "Transaction ID cannot be blank")
		Assert.notEmpty(event.tasks, "Tasks list cannot be empty")
		Assert.isTrue(event.playerId > 0, "Player ID must be positive")

		event.tasks.forEach {
			Assert.hasText(it.taskId, "Task ID cannot be blank")
		}
	}

	private fun sendSuccessNotification(event: SaveTasksEvent) {
		try {
			val sendNotificationEvent = SendNotificationEvent(
				event.transactionId,
				event.playerId,
				NotificationPriority.LOW,
				Notification(
					messageSource.getMessage(
						TASKS_GENERATION_SUCCESS,
						null,
						LocaleContextHolder.getLocale()
					),
					NotificationType.INFO,
					NotificationSource.TASKS
				)
			)

			sendNotificationProducer.send(sendNotificationEvent)
			log.info("Success notification sent for transaction {}", event.transactionId)

		} catch (e: Exception) {
			log.error(
				"Failed to send success notification for transaction {}: {}",
				event.transactionId,
				e.message
			)
		}
	}
}
